ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/DownloadSyncReceiptsStep.java

```diff
@@ -0,0 +1,98 @@
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.peertask.PeerTaskExecutorResponseCode;
+import org.hyperledger.besu.ethereum.eth.manager.peertask.PeerTaskExecutorResult;
+import org.hyperledger.besu.ethereum.eth.manager.peertask.task.GetReceiptsFromPeerTask;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+public class DownloadSyncReceiptsStep
+    implements Function<List<SyncBlock>, CompletableFuture<List<SyncBlockWithReceipts>>> {
+
+  private final ProtocolSchedule protocolSchedule;
+  private final EthContext ethContext;
+
+  public DownloadSyncReceiptsStep(
+      final ProtocolSchedule protocolSchedule, final EthContext ethContext) {
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+  }
+
+  @Override
+  public CompletableFuture<List<SyncBlockWithReceipts>> apply(final List<SyncBlock> blocks) {
+    final List<BlockHeader> headers = blocks.stream().map(SyncBlock::getHeader).collect(toList());
+    return ethContext
+        .getScheduler()
+        .scheduleServiceTask(() -> getReceiptsWithPeerTaskSystem(headers))
+        .thenApply((receipts) -> combineSyncBlocksAndReceipts(blocks, receipts));
+  }
+
+  private CompletableFuture<Map<BlockHeader, List<TransactionReceipt>>>
+      getReceiptsWithPeerTaskSystem(final List<BlockHeader> headers) {
+    Map<BlockHeader, List<TransactionReceipt>> getReceipts = new HashMap<>();
+    do {
+      GetReceiptsFromPeerTask task = new GetReceiptsFromPeerTask(headers, protocolSchedule);
+      PeerTaskExecutorResult<Map<BlockHeader, List<TransactionReceipt>>> getReceiptsResult =
+          ethContext.getPeerTaskExecutor().execute(task);
+      if (getReceiptsResult.responseCode() == PeerTaskExecutorResponseCode.SUCCESS
+          && getReceiptsResult.result().isPresent()) {
+        Map<BlockHeader, List<TransactionReceipt>> taskResult = getReceiptsResult.result().get();
+        taskResult
+            .keySet()
+            .forEach(
+                (blockHeader) ->
+                    getReceipts.merge(
+                        blockHeader,
+                        taskResult.get(blockHeader),
+                        (initialReceipts, newReceipts) -> {
+                          throw new IllegalStateException(
+                              "Unexpectedly got receipts for block header already populated!");
+                        }));
+        // remove all the headers we found receipts for
+        headers.removeAll(getReceipts.keySet());
+      }
+      // repeat until all headers have receipts
+    } while (!headers.isEmpty());
+    return CompletableFuture.completedFuture(getReceipts);
+  }
+
+  private List<SyncBlockWithReceipts> combineSyncBlocksAndReceipts(
```

Comment:
Think this method is the only change between this new class and `DownloadReceiptsStep`, could they be combined to remove duplication?

Reply:
done